### DISCUSSION TITLE
localization - injecting filtered data back to the Create 3

### DISCUSSION THREAD

**Discussion by arielc-brillianetor** [iRobotEducation/create3_docs#115] [Questions]
Hi,
I want to add localization abilities to my create3, but I want the robot to use the data which I calculate instead of only using the odometer and the imu installed on the robot. What I'm trying to do is use the move_to_position but with a more accurate result.
Is there a way to inject processed data back into the robot so it thinks that it is where I tell it instead of using only it's own localization process?
Thanks a lot for your help,
Ariel

p.s. not sure where this question is supposed to go, I apologize if it's in the wrong place.

**Comment by alsora** [comment]
HI @arielc-brillianetor,
currently there isn't a simple way to inject filtered data on the robot.
In general, what you are describing could be achieved by "remapping" ros2 topics.
So you would remap the existing odometry topic to a different name and take care of publishing on the original topic.

This is currently a feature that it's not exposed by create 3.
Note that, although it's easy in theory to add remapping arguments to the create 3 webserver,  there are some non-trivial considerations to take into account (i.e. it's likely that by blindly remapping topics you may bring the robot in a non-functional state, especially without knowing how the topics are used internally by the robot).

Remaps are on our to-do list, but I don't think we will release them in the near future.

My recommendation is that you should implement your own "move to position" behaviors.
These behaviors would be executed on your laptop or on you SBC (e.g. raspberry pi), they can use whatever input data you want and they will publish `Twist` messages on the `/cmd_vel` topic to drive the robot.
You can look at the open-source `create3_sim` repository for examples on how to implement behaviors similar to the ones that the robot is running under the hood (https://github.com/iRobotEducation/create3_sim/blob/main/irobot_create_common/irobot_create_nodes/src/motion_control/drive_goal_behaviors.cpp)

Alternatively you can use the ROS 2 Navigation stack.

**Reply by arielc-brillianetor** [reply]
What about the action move_distance? Is it more accurate then move_to_position? The problem is that I don't have an SBC to mount on the robots.
Thanks for your help
